### PR TITLE
Allow clouds to launch slaves with job-specified parameters

### PR DESCRIPTION
**Pull request by magnayn**
We would like to be able to add parameters into a build to contain
a setting for various slave parameters - for example, a docker volume
that contains a Jenkins workspace.

This would enable that workspace to be 'passed up' through a build pipeline
to subsequent stages.

However - with the current API, there is no way of the slave launcher
knowing what additional slave startup parameters that it needs. By the
time the slave is constructed, this is too late - such settings can only
be applied at slave construction time.

Thus there is a need for the slave launcher to be aware of (optional)
parameters that it can use in order to control the slave that it is
launching.

So -
- create a new CloudProvisioningRequest object, that contains the existing
  settings (Label, Excess Workload), and a list of 'Actionable' items that
  it is passed. Add a provision option into the Cloud class (default impl
  for backwards compatibility to be the old provision impl).

These Actionable items are, in effect, the builds in the queue. The cloud
subclass can examine this queue and pull off builds. From there it would
look for the ParametersAction, and the Parameters stored within. Cloud
implementations will provide their own ParameterDefinition/Values that
they look for.

Once a slave is launched, we need to make sure the 'right' job is attached
to the 'right' slave. Fortunately, all a cloud implementation needs to do
is add a NodeProperty to the launched slave, which Jenkins will ask
'canTake' to see if a buildable item is appropriate for this slave. All
the slave then needs to do is say it's not appropriate for that node if
the launch parameters don't match the build parameters.

Signed-off-by: Nigel Magnay nigel.magnay@gmail.com

### CODE REVIEW

**Comment by KostyaSha** [comment]
@stephenc could you review?

File: core/src/main/java/hudson/slaves/CloudProvisioningRequest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2010, InfraDNA, Inc.
```
Comment (KostyaSha) [inline core/src/main/java/hudson/slaves/CloudProvisioningRequest.java]: Seems not valid

File: core/src/main/java/hudson/slaves/CloudProvisioningRequest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   *      Number of total executors needed to meet the current demand.
+   *      Always >= 1. For example, if this is 3, the implementation
+   *      should launch 3 slaves with 1 executor each, or 1 slave with
+   *      3 executors, etc.
+   * @param queue
+   *      Collection of buildable items that are in the queue. Cloud implementations may
+   *      use this information to influence the kinds of nodes that they start up
+   *      (e.g: with specialised requirements based on the build).
+   */
+  CloudProvisioningRequest(Label label, int workloadToProvision,
```
Comment (KostyaSha) [inline core/src/main/java/hudson/slaves/CloudProvisioningRequest.java]: should it be public?

File: core/src/main/java/hudson/model/Queue.java [hunk 3]
```diff
@@ -839,6 +840,23 @@ public synchronized int countBuildableItemsFor(Label l) {
     }
 
     /**
+     * Get the {@link BuildableItem}s that are assigned to the given label
+     */
+    public synchronized Collection<Actionable> getBuildableItemsFor(Label l) {
+        Collection<Actionable> items = new LinkedHashSet<Actionable>();
+
+        for (BuildableItem bi : buildables.values())
```
Comment (lanwen) [inline core/src/main/java/hudson/model/Queue.java]: please add brackets

**Comment by stephenc** [comment]
The issue I see with this is that you need a way to flag that a `PlannedNode` is going to satisfy some of these `Actionable` items that are waiting for the label.

So while the idea of moving to a request object is a good pattern to follow, my belief is that this change as currently envisioned will not solve the problem you describe.

What we really need to have is some other kind of `Label`-like marker for tasks that can contain the extra information. Labels are really a specialized form of these markers and as such we should really be provisioning against markers and not against Labels as that would let us ensure that nodes are provisioned correctly.

The other thing is that I suspect your use of the current Cloud API is not aligned with its original vision.

The original vision of the Cloud API is that the future that gets wrapped in the PlannedNode represents all the work required to provision the node and configure it correctly. A simple launcher should be returned. What I see happening in most implementations of the Cloud API is that the future returns too early and then people use a custom launcher that "waits" for the node to be ready. As such the node customization should happen in the future so that the correct Node instance is constructed and returned *only when it is ready to be connected to*.

> Thus there is a need for the slave launcher to be aware of (optional)
parameters that it can use in order to control the slave that it is
launching.

is not really correct in that respect, because you should not be returning a custom slave launcher rather just one of the standard ones. The `Callable<Node>` that has been submitted to `Computer.threadPoolForRemoting` and is linked to the `PlannedNode.future` is what needs to be aware of (optional) parameters

**Comment by magnayn** [comment]
Bearing in mind most things Jenkins have been popped from my brain-stack..

What I really want is for the thing that launches the slave to be able to inspect the job it is being asked to run, and make decisions about how it launches that slave based on various strategies that it can determine. Currently there doesn't seem to be a way to do this, as the slave is both launched and the SSH connection established before there's an opportunity.

The simplest example is (docker)
- upstream job finishes a build.
- a task or teardown commits the data container, and pushes it to a docker registry
- the build is marked with the ID that is.

- downstream job is triggered
- as a part of the provisioning, the above is inspected, and the relevant data container downloaded and attached to the new build

But I can see more complex cases (e.g: downstream build walking a number of upstream builds to figure out what to do) and pluggable strategies. Adding "special labels" feels to like pre-deciding which bits of information may be useful rather than having access to everything in order to make that decision?

**Comment by stephenc** [comment]
What I'm saying is that Labels are a specific form of requirement. Currently they are the only form of requirement, so that is all we can work with.

What we actually want to do is provide a mechanism for Queue.Task to advertise its "node requirements" one of those may be a `Label` expression if the task is restricted to a specific label. The other ones will be the non-label requirements. This could be thinks like "based on a specific docker image" or "with the following tool installers" or "with the foo from bar".

When provisioning, you then gather up the set of "node requirements" and see, for each unique set, how many nodes there are and how many are available. Based on the node provisioning strategy you then decide to consult the `Cloud`s to see if any can provision a node meeting the specific set of "node requirements".

This current pull request makes the decision of canProvision based entirely on the Label. So what will happen is, if there are multiple clouds that can provision that label, but only the last one can provision a node with the other requirements, then the jobs that are blocked waiting for the other requirements will sit there until there are enough other jobs to exhaust the preceding cloud instances (and if one of those is an unlimited one, e.g. EC2 without a cost bound applied, well good luck).

So in a sense what we want is something more like

```
public boolean canProvision(Set<NodeRequirement> requirements) {
  // default impl for legacy implementations
  for (NodeRequirement r: requirements) {
    if (r instanceof LabelNodeRequirement) {
      if (!canProvision(LabelNodeRequirement.class.cast(r).getLabel())) {
        return false;
      }
    }
  }
  return true;
}
```

Now we could wrap `Set<NodeRequirement>` in a request object, but that really becomes more of a stylistic issue.

The current `NodeProvisioner` pattern of having one per label becomes problematic as really it should be one per unique `Set<NodeRequirement>` based on the tasks in the queue... with the left items being included so that you retain some of the history of requirements

**Comment by magnayn** [comment]
>
> What I'm saying is that Labels are a specific form of requirement.
> Currently they are the only form of requirement, so that is all we can work
> with.
>
> What we actually want to do is provide a mechanism for Queue.Task to
> advertise its "node requirements" one of those may be a Label expression
> if the task is restricted to a specific label. The other ones will be the
> non-label requirements. This could be thinks like "based on a specific
> docker image" or "with the following tool installers" or "with the foo from
> bar".
>
​I guess as long as there's an extensible mechanism for plugins to be able
to add these other requirements in based on inspecting other builds

When provisioning, you then gather up the set of "node requirements" and
> see, for each unique set, how many nodes there are and how many are
> available. Based on the node provisioning strategy you then decide to
> consult the Clouds to see if any can provision a node meeting the
> specific set of "node requirements".
>
> This current pull request makes the decision of canProvision based
> entirely on the Label. So what will happen is, if there are multiple clouds
> that can provision that label, but only the last one can provision a node
> with the other requirements, then the jobs that are blocked waiting for the
> other requirements will sit there until there are enough other jobs to
> exhaust the preceding cloud instances (and if one of those is an unlimited
> one, e.g. EC2 without a cost bound applied, well good luck).
>
​Ah - in my usecases I'd never have the scenario where a cloud (provided
the labels matched) was ever unable to ​satisfy the additional
requirements, they're pretty much container launch parameters.

There might be better choices than others (e.g: building on the same
physical node as the upstream build).

**Comment by KostyaSha** [comment]
In other build systems its called "constraints" where you can define os, memory limit and etc and then scheduler/dispatcher decides what to do.
Imho both suggestions are valid.
Label itself can contain any information, so Label can also be a valid object for containing such info, but then it should be wrapped to some Set instead of single object. 
@stephenc variant with having new class and hierarchy is also good because it will provide extensibility. For example memory constraint may work then as comparison, while label now can do only strict equals. 

Imho it probably also makes sense to remove all custom implementation of canProvision() overrides and provide some mechanism of contributing available info from clouds and then define on API side whether to request provisioning from Cloud. For example, cloud says that it supports mode.NORMAL and 5 containers with 5gb ram and etc, then some scheduler/dispatcher calls some contributed mechanisms and decides that this node is OK for provisioning.  (My idea is remove custom label comparison + mode.* calculations from Clouds that implemented almost everywhere differently).

**Comment by stephenc** [comment]
Well you know my view that the current Cloud API is a pile of shite and none have implemented it correctly anywhere

(I tag my own implementations in the "pile of shite" bag too... they only work because they have a layer above that fixes the problems induced by the Cloud API... so do not feel bad if you have written a Cloud API implementation, I am of the opinion that it is impossible to write a completely correct implementation of the current Cloud API)

**Comment by KostyaSha** [comment]
How that correlates with https://wiki.jenkins-ci.org/display/JENKINS/Compound+Slaves ?
According to JUC report, people wanted the same but decided to not deal with core. How much should everybody wait for good provisioning? :disappointed: 
@stephenc do you need any help? I can join development if you have some plans.

**Comment by superbob** [comment]
Coming from https://github.com/jenkinsci/docker-plugin/issues/74
+1 on this one also, is there any chance to have it included?

**Comment by magnayn** [comment]
jenkinsci/docker-plugin#74
<https://github.com/jenkinsci/docker-plugin/issues/74> may or may not
require this, depending on your interpretation of that ticket!

Some people would like to launch docker slaves, and be able to pass them
--link parameters. This is relatively simple, is just waiting for someone
to care enough to contribute it, and doesn't need this ticket.

Some people would like to launch numbers of containers together, link them
together, and then tear them down together - a bit like docker-compose.
This is harder, but still not too hard, and doesn't need this ticket.

Still further people (like, me) want(ed) the ability for slaves to launch
with knowledge about the build they were being launched for to perform. So
when an upstream build finishes, it could commit the docker container, and
a downstream build recommence where the upstream one left off - or, to
encapsulate the workspace in a docker data container, or... whatever.
Basically prevent a lot of up/downstream artifact pushing and pulling.
Sadly once the docker container is launched, you can't add links, so it's
too late. This ticket was one suggetion that might make it work.

It's been stalled for a year (terminating in largely the belif this PR
won't work, something different needed, cloud API not very good) so it
should probably just be closed; I've largely solved my usecases with
jenkins-workflow so my needs are less pressing.

**Comment by superbob** [comment]
I was also considering jenkins-workflow, but that seemed to be "too much" for my use case. We are already using "standard" jenkins jobs everywhere, and going into workflow would be a major change in our uses. Basically, I just need the --links (https://github.com/jenkinsci/docker-plugin/issues/74). I'ld also like to be able to override my docker templates parameters in my job, because going to jenkins system settings all the time is really a hassle, but that'd be just a bonus.

I've understood there are issues with cloud API. Do we know where this is going? Will jenkins-docker continue to use it? Are there plans to do it 'right'? (Just throwing these away, don't hope to have a lot of answers...)

File: core/src/main/java/hudson/slaves/Cloud.java [hunk 4]
```diff
@@ -116,6 +116,21 @@ public final boolean hasPermission(Permission permission) {
...
+    {
+        // Non-abstract default impl, so newer clients don't have to implement
+        // a deprecated method.
+        throw new AbstractMethodError("You must override a provision method");
+    }
+
+    /**
+     * Returns true if this cloud is capable of provisioning new nodes for the given label.
+     */
+    public abstract boolean canProvision(Label label);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/Cloud.java]: Not compatible, default impl is a must

**Comment by daniel-beck** [comment]
@ndeloof How does this relate to your one-shot executors?

**Comment by KostyaSha** [comment]
@daniel-beck orthogonal, this PR uses existing way of provisioning slaves.

**Comment by ndeloof** [comment]
@daniel-beck this indeed seems to address a very similar goal (haven't looked in details, just read description), but doesn't include the idea to make slave dedicated to a specific build, i.e. won't be reused, and should fail the build if it can't provision with configured parameters.

**Comment by olivergondza** [comment]
I tend to agree as such functionality now exists in declarative pipeline.

**Comment by ndeloof** [comment]
for the record, this exact use case was a trigger for me to create one-shot-executor-plugin, but I never have been able to make this a non-hack-ish solution
